Indicate search results are limited to the sample size in a saved search on dashboard

stacey-gammon opened the pull request:
<img width="932" alt="screen shot 2017-03-21 at 8 46 12 am" src="https://cloud.githubusercontent.com/assets/16563603/24147979/26aa2898-0e13-11e7-95cd-516cfc24f45a.png">


Fixes #4161

Stacey-Gammon commented:
Jenkins failure: (`Fatal error: unexpected end of file`)
```
Running "esvm:test" (esvm) task
starting up "test" cluster
INFO -  - cluster - Downloading & installing from "master" branch.
Fatal error: unexpected end of file
runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
runbld>>> DURATION: 234394ms
runbld>>> STDOUT: 81598 bytes
runbld>>> STDERR: 2050 bytes
```

jenkins, test this

spalger commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -20,6 +20,12 @@
       </tbody>
     </table>
+    <div ng-if="hits.length === sampleSize && page.last" class="discover-table-footer">
+      <center>
```
@cjcenizal I'm tempted to say we shouldn't use `<center>` tags, do you agree?

cjcenizal replied:
Ah, yes definitely. Centering should be done in CSS, preferably with flexbox, but using `text-align: center` is also an option. In this case, I'd check the `.discover-table-footer` class to see if it makes sense to add these styles to that class. If there are situations where we don't want to center it's content, then create a new class, e.g.`.discover-table-footer-prompt`, and apply the style there. The updated markup would look like this:

```html
<div ng-if="hits.length === sampleSize && page.last" class="discover-table-footer">
  <div class="discover-table-footer-prompt">
    These are the first {{sampleSize}} documents matching
    your search, refine your search to see others.
  </div>
</div>
```

spalger commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -20,6 +20,12 @@
           class="discover-table-row"></tr>
       </tbody>
     </table>
+    <div ng-if="hits.length === sampleSize && page.last" class="discover-table-footer">
```
In discover, [`$scope.hits`](http://github.com/elastic/kibana/blob/13ebe995cd2b8624fe86ede612148b4579b4520a/src/core_plugins/kibana/public/discover/controllers/discover.js#L422) is the total number of results. Maybe we should pass that to the doc_table and it should use that to determine if this should be shown (which means it could also tell the user how many rows are hidden).

Stacey-Gammon replied:
We can't tell the user how many rows are hidden when using this from dashboard because we don't have the total number of results in that case, we are only passing searchSource.

Discover won't ever show this part because it uses infiniteScroll and passes hits instead of using searchSource & setting hits.  I agree it's unintuitive that `hits` is two different things depending on infiniteScroll and searchSource, but I don't want to get into a refactor since this is a Mend it Monday `low fruit` issue and if I make larger modifications it should probably go through testing.  Thoughts?

Stacey-Gammon replied:
Oh the other thing is that in the discover footer now, there is a `back to top` link where the logic is in discover, not in doc table.  I got rid of that, since it doesn't make sense esp if this error message goes on top, not on the bottom.

spalger replied:
Forgot that this is used outside of discover

jimmyjones2 commented:
The other issue in #4161 was that on the first page you don't know that the results have been limited. Could there be some kind of indication on the pager perhaps, or at the top of the first page?

Stacey-Gammon commented:
Took into account @jimmyjones2 feedback and moved the warning back to the first page, on top.  Split out the styles since it's no longer a footer, and was never part of discover anyhow.  And got rid of the center tag.

<img width="920" alt="screen shot 2017-03-27 at 10 11 09 am" src="https://cloud.githubusercontent.com/assets/16563603/24360650/ee1ff06c-12d5-11e7-92d8-34ace506c261.png">

@cjcenizal @snide I went back to the first page, on top ^^

trevan commented:
@Stacey-Gammon, one minor problem I have is that it takes up a lot of space on the first page.  I have several dashboards that have multiple discover searches on it.  I'd like to see as much data as possible but now with this warning and with the pagination, I see a lot less than I did which requires either scrolling or increasing the panel sizes.

jimmyjones2 commented:
Could the warning be inline with the pager, so doesn't consume vertical space? Would just "limited to first 500 results" suffice?

Stacey-Gammon commented:
Using the new pager styles:
<img width="1288" alt="screen shot 2017-03-29 at 3 43 43 pm" src="https://cloud.githubusercontent.com/assets/16563603/24473416/319faaf4-1497-11e7-8df4-17fe4c32e4e3.png">

cjcenizal commented:
Hmm, I think it does look better on the right side. I think we can declutter the UI by hiding the warning message behind a tooltip. @snide @uboness thoughts?

![image](https://cloud.githubusercontent.com/assets/1238659/24474086/5e4fd950-1480-11e7-93fd-6476105d5b62.png)

![image](https://cloud.githubusercontent.com/assets/1238659/24474091/60248618-1480-11e7-8263-2a54587e301c.png)

@Stacey-Gammon If we go this route, we'll need to add a custom class to the Bar element here, which adds `margin: 5px 5px 0` to adjust the spacing.

uboness commented:
+1 on right alignment 

regarding the message, not sure if the tooltip will be too hidden. We could have a "calmer" text showing closer to the nav control... not sure... need to see it in action.. might be that a tooltip is good enough

snide commented:
Right alignment is good. I'd use whatever our light gray text color is and make the text more succinct to keep it from being so heavy. "Limited to 500 results. Refine your search." should get the point across.

Stacey-Gammon commented:
@snide - Good suggestions, I'll add them shortly.
@cjcenizal - Note that the warning message only appears on the very last page so I don't think it'll be too intrusive (or at least that's how I have it implemented now). It will only show up after the user pages through 10 times, so I think we want it a little more obvious than the icon (IMO).

uboness commented:
> Note that the warning message only appears on the very last page so I don't think it'll be too intrusive (or at least that's how I have it implemented now). It will only show up after the user pages through 10 times, so I think we want it a little more obvious than the icon (IMO).

++

> 'd use whatever our light gray text color is and make the text more succinct to keep it from being so heavy. "Limited to 500 results. Refine your search." should get the point across.

++

cjcenizal commented:
@snide Good suggestions! Let's do it.

Stacey-Gammon commented:
Latest iteration:

<img width="1115" alt="screen shot 2017-03-30 at 2 11 36 pm" src="https://cloud.githubusercontent.com/assets/16563603/24519420/1636edc4-1553-11e7-96b1-27ca64ae67fe.png">

cjcenizal commented:
@Stacey-Gammon Looks good, but the order should go [Limited message] [Result count] [Buttons], to retain as close a pattern to what we currently use:

![image](https://cloud.githubusercontent.com/assets/1238659/24519823/20dd3b60-153b-11e7-91bc-c3217df11358.png)

Stacey-Gammon commented:
Oh good call, forgot I swapped them earlier.  This looks better.

<img width="1050" alt="screen shot 2017-03-30 at 2 26 25 pm" src="https://cloud.githubusercontent.com/assets/16563603/24520054/0a111824-1555-11e7-98b2-5e657ae417ea.png">

Stacey-Gammon commented:
jenkins test this

Stacey-Gammon commented:
@cjcenizal Does `looks good` refer only to UI/UX or did you check out the code as well?

cjcenizal commented:
Just the screenshot. I can do a code review too if you want.

Stacey-Gammon commented:
@cjcenizal I do need a second reviewer so that would be great, but can tag someone else if you are busy!

cjcenizal commented on src/ui/public/doc_table/doc_table.js:
```diff
@@ -102,13 +104,38 @@ uiModules.get('kibana')
       }));
 
+      const limitTo = $filter('limitTo');
+      const calculateItemsOnPage = () => {
```
How about moving `limitTo` and `calculateItemsOnPage` up to line 53 where the other non-scope variables are? This way the non-scope and scope variables/methods will be grouped together, and you can remove the `eslint-disable-line` annotation on 111.

cjcenizal commented on src/ui/public/doc_table/doc_table.less:
```diff
@@ -8,6 +8,16 @@ doc-table {
+    background-color: @doc-table-warning-bg;
+    padding: 5px 10px;
+    text-align: center;
+  }
```
Is this class being used anywhere?

cjcenizal commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -1,8 +1,28 @@
+  ng-class="{ loading: searchSource.activeFetchCount > 0 }"
+>
+  <div ng-if="!infiniteScroll">
+    <div class="kuiBar">
```
I updated UI Framework with a new treatment for making the text gray. Can you rebase and update the markup to this:

```html
    <div class="kuiBar docTableBar">
      <div class="kuiBarSection">
        <div
          ng-if="shouldShowLimitedResultsWarning()"
          class="kuiToolBarText kuiSubduedText">
          {{ limitedResultsWarning }}
        </div>
```

cjcenizal commented on src/ui/public/doc_table/doc_table.less:
```diff
@@ -8,6 +8,16 @@ doc-table {
   flex: 1 1 100%;
   flex-direction: column; /* 1 */
 
+  .docTablePagerInfoMessage {
```
Can you delete this class and add this new one instead? And can you unnest it and put it on line 3, outside of `doc-table`?

```less
.docTableBar {
  margin: 5px 5px 0;
}
```

cjcenizal commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -20,7 +40,27 @@
     </table>
-  </paginate>
+    <!-- ToolBarFooter -->
+    <div class="kuilBar">
```
And this can be updated too:

```html
    <div class="kuilBar docTableBar">
      <div class="kuiBarSection">
        <div
          ng-if="shouldShowLimitedResultsWarning()"
          class="kuiToolBarText kuiSubduedText">
          {{ limitedResultsWarning }}
        </div>
```

cjcenizal commented on src/ui/public/styles/variables/for-theme.less:
```diff
@@ -41,6 +41,7 @@
 
 @discover-table-footer-bg: @well-bg;
 @discover-field-filter-bg: @well-bg;
+@doc-table-warning-bg: @well-bg;
```
You can delete this.

cjcenizal commented on src/ui/public/filter_bar/filter_bar.js:
```diff
@@ -46,10 +45,6 @@ module.directive('filterBar', function (Private, Promise, getAppState) {
 
-      // Don't show filter "pinnability" when in embedded mode, as it doesn't make sense in that context
-      // as there will be no cross app navigation for which the filter should persist.
-      $scope.showFilterPin = () => chrome.getVisible();
```
Can you explain the context behind these changes? At first glance they don't seem related to the other changes in this PR.

Stacey-Gammon replied:
Weird.  I had another PR that accidentally got merged into this one, so I reverted it, hence the commit history on this one was a little weird: 1) merge unrelated PR. 2) Do actual work. 3) Notice mistake and revert PR.

But now that original PR is merged for real, and I guess the revert is making this show up as actually removing that code.

Anyway, good catch, I'll add it back in.

Stacey-Gammon commented:
@cjcenizal Feedback addressed, lmk if there is anything else!

cjcenizal commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -1,8 +1,30 @@
+      <div class="kuiBarSection">
+        <div
+          ng-if="shouldShowLimitedResultsWarning()"
+          class="kuiToolBarText kuiSubduedText">
```
This closing bracket should be on the next line.

cjcenizal commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -23,7 +45,29 @@
+      <div class="kuiBarSection">
+        <div
+          ng-if="shouldShowLimitedResultsWarning()"
+          class="kuiToolBarText kuiSubduedText">
```
Same here.

cjcenizal commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -1,8 +1,30 @@
+  ng-class="{ loading: searchSource.activeFetchCount > 0 }"
+>
+  <div ng-if="!infiniteScroll">
+    <div class="kuilBar docTableBar">
```
Looks like a typo here: `kuilBar` should be `kuiBar`.

cjcenizal commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -23,7 +45,29 @@
     </table>
-  </paginate>
+    <!-- ToolBarFooter -->
+    <div class="kuilBar docTableBar">
```
`kuilBar` should be `kuiBar`.
